network/src/noise_wrapper/handshake.rs

```diff
@@ -0,0 +1,348 @@
+//   in order to pass them to the noise implementaiton
+//
+
+/// The Noise configuration to be used to perform a protocol upgrade on an underlying socket.
+pub struct NoiseWrapper(noise::NoiseConfig);
+
+impl NoiseWrapper {
+    /// Create a new NoiseConfig with the provided keypair
+    pub fn new(key: x25519::PrivateKey) -> Self {
+        Self(noise::NoiseConfig::new(key))
+    }
+
+    /// Perform a protocol upgrade on an underlying connection. In addition perform the noise IX
+    /// handshake to establish a noise session and exchange static public keys. Upon success,
+    /// returns the static public key of the remote as well as a NoiseSession.
+    // TODO(mimoo, philp9): this code could be inlined in transport.rs once the monolithic network is done
+    pub async fn upgrade_connection<TSocket>(
+        &self,
+        socket: TSocket,
+        origin: ConnectionOrigin,
+        anti_replay_timestamps: Option<Arc<RwLock<AntiReplayTimestamps>>>,
+        remote_public_key: Option<x25519::PublicKey>,
+        trusted_peers: Option<&Arc<RwLock<HashMap<PeerId, NetworkPeerInfo>>>>,
+    ) -> io::Result<(x25519::PublicKey, NoiseSession<TSocket>)>
+    where
+        TSocket: AsyncRead + AsyncWrite + Unpin,
+    {
+        // perform the noise handshake
+        let socket = match origin {
+            ConnectionOrigin::Outbound => {
+                let remote_public_key = match remote_public_key {
+                    Some(key) => key,
+                    None if cfg!(any(test, feature = "fuzzing")) => unreachable!(),
+                    None => {
+                        return Err(std::io::Error::new(
+                            std::io::ErrorKind::Other,
+                            "noise: SHOULD NOT HAPPEN: missing server's key when dialing",
+                        ));
+                    }
+                };
+                self.dial(socket, anti_replay_timestamps.is_some(), remote_public_key)
+                    .await?
+            }
+            ConnectionOrigin::Inbound => {
+                self.accept(socket, anti_replay_timestamps, trusted_peers)
+                    .await?
+            }
+        };
+
+        // return remote public key with a socket including the noise session
+        let remote_public_key = socket.get_remote_static();
+        Ok((remote_public_key, socket))
+    }
+
+    pub async fn dial<TSocket>(
+        &self,
+        mut socket: TSocket,
+        validator_network: bool,
+        remote_public_key: x25519::PublicKey,
+    ) -> io::Result<NoiseSession<TSocket>>
+    where
+        TSocket: AsyncRead + AsyncWrite + Unpin,
+    {
+        // on the validator network, send prologue as current timestamp (in seconds)
+        let mut prologue = [0u8; 8];
+        if validator_network {
+            let now: u64 = time::SystemTime::now()
+                .duration_since(time::UNIX_EPOCH)
+                .expect("system clock should work")
+                .as_secs();
```

Comment:
I would rather us be at millisecond granularity at the network layer and set us up for being able to tune it later, we don't see it often now, but we may see it later.  Most networking happens on the order of milliseconds between datacenters.  What's the cost implication of trying again?  I'm still catching up a little on what this exactly does :)

Reply:
i'd rather we don't conflate rate limiting with replay attack detection.

instead of importing another crate, let's just do:

```rust
let now: u128 = time::SystemTime::now()
    .duration_since(time::UNIX_EPOCH)
    .expect("system clock should work")
    .as_nanos();
let payload: [u8; 16] = now.to_le_bytes();
```

let's also send the timestamp in the encrypted payload, since we're not actually using it until we parse the whole client init message. this would also hide highly structured data (nanos timestamp) from being visible over-the-wire.

Reply:
why nanos though? If I understand the dailer correctly it's only going to take milliseconds to have another dial right?